### PR TITLE
Add view renderer superclass for easily creating WAF components

### PR DESCRIPTION
**Pull request by kodikos**
This adds a class for creating renderers that have the same features as the HTML renderer but output in the WAF JSON format. You can define 2 special methods, `head` and `body_last` to place items into the top level head and bodyLast keys of the JSON output.

kungfu_wafers

### Example

These files provide a really simple example of usage:
`src/components/wafs/models/wafer.rb`
```ruby
require 'alephant/renderer/views/waf'
module DataRenderer
  module Views
    class Wafer < Alephant::Renderer::Views::Waf
      def mood
        t 'jolly'
      end
    end
  end
end
```
`src/components/wafs/templates/wafer.mustache`
```html
<h1>Hello, I'm a {{mood}} template</h1>
```
`src/components/locale/en.yml`
```yaml
  wafer:
    jolly: 'jolly'
```
Will output:
```json
{"head":[],"bodyInline":"<h1>Hello, I'm a jolly template</h1>\n","bodyLast":[]}
```

### CODE REVIEW

**Comment by kodikos** [comment]
@tnyuan, @samfrench, @dazoakley, @ettomatic, @zygotic99.. for review. I doubt I have time to actually fix anything, one of you will have to move this through.. will take Tian through it before I go.

**Comment by samfrench** [comment]
👍  Looks fine to me.

The only downside of this, would be having to create a separate template for waf and html. This would be the equivalent of en or cy, but would be expected.

**Comment by samfrench** [comment]
Is envelope a better name than waf? Just thinking the format of the response is an envelope.

**Comment by samfrench** [comment]
Second point, not sure if the requirement is to render a component in the waf format and html simultaneously without creating two templates? That would be the simplest thing, but not sure on the exact requirement.

**Comment by Tarqwyn** [comment]
I think maybe you are right about Waf versus Envelope.. gonna change it

File: spec/fixtures/components/foo-renderer/models/waf.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+        def body_last
+          ['some/thing/to/add/after/body', 'some/other/thing/to/add/after/body']
+        end
+
+        def greet
+          @data[:foo]
+        end
+
+  end
```
Comment (dazoakley) [inline spec/fixtures/components/foo-renderer/models/waf.rb]: The spacing/formatting is a bit off here...

**Comment by dazoakley** [comment]
Other than the formatting in that one file, looks good to me.

I don't mind on the `Waf` v's `Envelope` name - sounds like you'd prefer `Envelope` so let's go with that. 👍

**Comment by Tarqwyn** [comment]
erm.. not sure of an easy way to render in two formats with one template automatically, though it could be done by pointing your `Envelope` model to your HTML template so both models use the same template. This would be OK for components that were for Mozart and the Pal as you would assume the `css` and `js` would be deployed on the pal as well..

I think a better solution is to keep the Renderer simple and if `Envelope` content needs to be available as just HTML then a two template approach is fine.. if it get cumbersome then perhaps some kind of proxy micro-service that takes `Envelopes` and serves them as HTML would be better keeping the renderer dum

**Comment by dazoakley** [comment]
You can share templates between models (at least between the ones that inherit from `Html` - all you need to do is override the `template_name` method.

**Comment by Tarqwyn** [comment]
thats what I thought.. there are solutions available to the dev... I think this approach keeps it simple

**Comment by kodikos** [comment]
Hi.. sorry I'm late to the convo! :wave:
re: naming - I'd used WAF because it signified the BBC-specific format we needed. Envelope is a bit more abstract to a casual user. @zygotic99 may want to comment on this, as I thought we'd agreed the naming together.
re: multi-output formats. Although nice in an ideal world, this isn't the use case for this. It's main aim was to make the WAF rendering compatible with HTML rendering for consistency and so we could more easily switch existing stuff to WAF. To render multiple formats is a non-trivial exercise, as you'd most likely need to involve broker changes. When we write renderers, we are pretty specific about the format we're deploying them in. At worst you can create a new renderer in the other format and instance the other renderer and call its' render method.

**Pull request merged**